### PR TITLE
Use SQL for storing data

### PR DESCRIPTION
**Pull request by DavidGlitch04**
Try to solve #10 issue

### CODE REVIEW

**Comment by NhanAZ** [comment]
Some lints, better if it was solved.
https://poggit.pmmp.io/ci/thebigcrafter/KygekTagsShop/~/pr:10

**Comment by DavidGlitch04** [comment]
All poggit lints have been fixed

File: src/Kygekraqmak/KygekTagsShop/TagsActions.php [hunk 1]
```diff
@@ -256,9 +275,14 @@ private function getData(Player $player) : int {
      * @param int $tagid
      */
     private function setData(Player $player, int $tagid) {
-        $this->data->set(strtolower($player->getName()), $tagid);
-        $this->saveData();
-        $this->reloadData();
+        //$this->data->set(strtolower($player->getName()), $tagid);
```
Comment (NhanAZ) [inline src/Kygekraqmak/KygekTagsShop/TagsActions.php:278]: Consider removing unless comment.

File: src/Kygekraqmak/KygekTagsShop/TagsActions.php [hunk 2]
```diff
@@ -256,9 +275,14 @@ private function getData(Player $player) : int {
...
+        $this->data->executeChange('kygektagsshop.update', [
+            'player' => strtolower($player->getName()),
+            'tagid' => $tagid
+        ]);
+        $this->data->waitAll();
+        //$this->saveData();
+        //$this->reloadData();
```
Comment (NhanAZ) [inline src/Kygekraqmak/KygekTagsShop/TagsActions.php:285]: Consider removing unless comment.

File: src/Kygekraqmak/KygekTagsShop/TagsActions.php [hunk 3]
```diff
@@ -268,37 +292,45 @@ private function setData(Player $player, int $tagid) {
...
     private function removeData(Player $player) {
-        $this->data->remove(strtolower($player->getName()));
-        $this->saveData();
-        $this->reloadData();
+        $this->data->executeChange('kygektagsshop.remove', [
+            'player' => strtolower($player->getName())
+        ]);
+        $this->data->waitAll();
+        //$this->saveData();
+        //$this->reloadData();
```
Comment (NhanAZ) [inline src/Kygekraqmak/KygekTagsShop/TagsActions.php:300]: Consider removing unless comment.

File: src/Kygekraqmak/KygekTagsShop/TagsActions.php [hunk 4]
```diff
@@ -268,37 +292,45 @@ private function setData(Player $player, int $tagid) {
...
 
     /**
      * Reloads the KygekTagsShop database
      */
-    public function reloadData() {
-        $this->data->reload();
-    }
+    // public function reloadData() {
+    //     $this->data->reload();
+    // }
```
Comment (NhanAZ) [inline src/Kygekraqmak/KygekTagsShop/TagsActions.php:325]: Consider removing unless comment.

File: src/Kygekraqmak/KygekTagsShop/TagsActions.php [hunk 5]
```diff
@@ -268,37 +292,45 @@ private function setData(Player $player, int $tagid) {
...
 
     /**
      * Saves the KygekTagsShop database
      */
-    public function saveData() {
-        $this->data->save();
-    }
+    // public function saveData() {
+    //     $this->data->save();
+    // }
```
Comment (NhanAZ) [inline src/Kygekraqmak/KygekTagsShop/TagsActions.php:333]: Consider removing unless comment.

File: src/Kygekraqmak/KygekTagsShop/TagsShop.php [hunk 6]
```diff
@@ -81,7 +84,12 @@ protected function onEnable(): void{
 
         $this->saveResource("config.yml");
         $this->config = $this->getConfig()->getAll();
-        $this->data = new Config($this->getDataFolder()."data.json", Config::JSON);
+        // $this->data = new Config($this->getDataFolder()."data.json", Config::JSON);
```
Comment (NhanAZ) [inline src/Kygekraqmak/KygekTagsShop/TagsShop.php:87]: Consider removing unless comment.

**Review by NhanAZ — COMMENTED** [review]
Please consider removing unless comment

**Comment by NhanAZ** [comment]
https://crash.pmmp.io/view/7289715

**Comment by NhanAZ** [comment]
# Error information
Message:	Undefined array key 0
File:	plugins/KygekTagsShop/src/Kygekraqmak/KygekTagsShop/TagsActions
Line:	#314
Error level:	ErrorException

```
#0 plugins/KygekTagsShop/src/Kygekraqmak/KygekTagsShop/TagsActions(314): pocketmine\errorhandler\ErrorToExceptionHandler::handle(int 2, string[21] Undefined array key 0, string[74] D:\Dev\plugins\KygekTagsShop\src\Kygekraqmak\KygekTagsShop\TagsActions.php, int 314)
#1 D:/Dev/virions/libasynql/src/poggit/libasynql/base/DataConnectorImpl(168): Kygekraqmak\KygekTagsShop\TagsActions->Kygekraqmak\KygekTagsShop\{closure}(array[0], array[0])
#2 D:/Dev/virions/libasynql/src/poggit/libasynql/base/DataConnectorImpl(183): poggit\libasynql\base\DataConnectorImpl::poggit\libasynql\base\{closure}(object poggit\libasynql\result\SqlSelectResult#119972)
#3 D:/Dev/virions/libasynql/src/poggit/libasynql/base/DataConnectorImpl(213): poggit\libasynql\base\DataConnectorImpl::poggit\libasynql\base\{closure}(array[1])
#4 D:/Dev/virions/libasynql/src/poggit/libasynql/base/SqlThreadPool(112): poggit\libasynql\base\DataConnectorImpl->poggit\libasynql\base\{closure}(array[1])
#5 D:/Dev/virions/libasynql/src/poggit/libasynql/base/DataConnectorImpl(296): poggit\libasynql\base\SqlThreadPool->readResults(array[1])
#6 D:/Dev/virions/libasynql/src/poggit/libasynql/base/SqlThreadPool(65): poggit\libasynql\base\DataConnectorImpl->checkResults()
#7 pmsrc/vendor/pocketmine/snooze/src/SleeperHandler(123): poggit\libasynql\base\SqlThreadPool->poggit\libasynql\base\{closure}()
#8 (): [38;5;227m--- below is the original stack trace ---[m()
#9 D:/Dev/virions/libasynql/src/poggit/libasynql/base/DataConnectorImpl(202): poggit\libasynql\base\DataConnectorImpl->executeImplRaw(array[1], array[1], int 3, object Closure#120027, null)
#10 D:/Dev/virions/libasynql/src/poggit/libasynql/base/DataConnectorImpl(182): poggit\libasynql\base\DataConnectorImpl->executeImpl(string[20] kygektagsshop.getall, array[0], int 3, object Closure#120027, null)
#11 D:/Dev/virions/libasynql/src/poggit/libasynql/base/DataConnectorImpl(166): poggit\libasynql\base\DataConnectorImpl->executeImplLast(string[20] kygektagsshop.getall, array[0], int 3, object Closure#120026, null)
#12 plugins/KygekTagsShop/src/Kygekraqmak/KygekTagsShop/TagsActions(315): poggit\libasynql\base\DataConnectorImpl->executeSelect(string[20] kygektagsshop.getall, array[0], object Closure#120025)
#13 plugins/KygekTagsShop/src/Kygekraqmak/KygekTagsShop/TagsActions(139): Kygekraqmak\KygekTagsShop\TagsActions->getAllData(object Closure#120024)
#14 plugins/KygekTagsShop/src/Kygekraqmak/KygekTagsShop/TagsShop(175): Kygekraqmak\KygekTagsShop\TagsActions->playerHasTag(object pocketmine\player\Player#30615)
#15 pmsrc/src/event/RegisteredListener(60): Kygekraqmak\KygekTagsShop\TagsShop->onJoin(object pocketmine\event\player\PlayerJoinEvent#119970)
#16 pmsrc/src/event/Event(62): pocketmine\event\RegisteredListener->callEvent(object pocketmine\event\player\PlayerJoinEvent#119970)
#17 pmsrc/src/player/Player(834): pocketmine\event\Event->call()
#18 pmsrc/src/network/mcpe/NetworkSession(710): pocketmine\player\Player->doFirstSpawn()
#19 pmsrc/src/network/mcpe/NetworkSession(703): pocketmine\network\mcpe\NetworkSession->onClientSpawnResponse()
#20 pmsrc/src/network/mcpe/handler/SpawnResponsePacketHandler(35): pocketmine\network\mcpe\NetworkSession->pocketmine\network\mcpe\{closure}()
#21 pmsrc/vendor/pocketmine/bedrock-protocol/src/SetLocalPlayerAsInitializedPacket(42): pocketmine\network\mcpe\handler\SpawnResponsePacketHandler->handleSetLocalPlayerAsInitialized(object pocketmine\network\mcpe\protocol\SetLocalPlayerAsInitializedPacket#120004)
#22 pmsrc/src/network/mcpe/NetworkSession(394): pocketmine\network\mcpe\protocol\SetLocalPlayerAsInitializedPacket->handle(object pocketmine\network\mcpe\handler\SpawnResponsePacketHandler#100570)
#23 pmsrc/src/network/mcpe/NetworkSession(357): pocketmine\network\mcpe\NetworkSession->handleDataPacket(object pocketmine\network\mcpe\protocol\SetLocalPlayerAsInitializedPacket#120004, string[2] q.)
#24 pmsrc/src/network/mcpe/raklib/RakLibInterface(182): pocketmine\network\mcpe\NetworkSession->handleEncoded(string[47] c*d4.......N.4..t....!......@.....LLLF ....{A4.)
#25 pmsrc/vendor/pocketmine/raklib-ipc/src/RakLibToUserThreadMessageReceiver(42): pocketmine\network\mcpe\raklib\RakLibInterface->onPacketReceive(int 0, string[56] .5-..].Y^+}Xt..U|"....w.~I..7....;..&aa{V.(......KV...(V)
#26 pmsrc/src/network/mcpe/raklib/RakLibInterface(112): raklib\server\ipc\RakLibToUserThreadMessageReceiver->handle(object pocketmine\network\mcpe\raklib\RakLibInterface#29000)
#27 pmsrc/vendor/pocketmine/snooze/src/SleeperHandler(123): pocketmine\network\mcpe\raklib\RakLibInterface->pocketmine\network\mcpe\raklib\{closure}()
#28 pmsrc/vendor/pocketmine/snooze/src/SleeperHandler(82): pocketmine\snooze\SleeperHandler->processNotifications()
#29 pmsrc/src/Server(1703): pocketmine\snooze\SleeperHandler->sleepUntil(float 1659508066.5728)
#30 pmsrc/src/Server(1063): pocketmine\Server->tickProcessor()
#31 pmsrc/src/PocketMine(304): pocketmine\Server->__construct(object BaseClassLoader#2, object pocketmine\utils\MainLogger#3, string[7] D:\Dev\, string[15] D:\Dev\plugins\)
#32 pmsrc/src/PocketMine(327): pocketmine\server()
#33 pmsrc(11): require(string[51] phar://D:/Dev/PocketMine-MP.phar/src/PocketMine.php)
```

> https://crash.pmmp.io/view/7289715

**Comment by DavidGlitch04** [comment]
Try again

**Review by NhanAZ — COMMENTED** [review]
Poggit Lints
https://poggit.pmmp.io/ci/thebigcrafter/KygekTagsShop/~/pr:25

**Comment by DavidGlitch04** [comment]
Retry Please

**Review by NhanAZ — COMMENTED** [review]
Consider using `$this` instead of `self::`

**Comment by NhanAZ** [comment]
```yaml
[11:28:22.836] [Server thread/DEBUG]: [KygekTagsShop] Queuing mode-3 query: SELECT * FROM kygektagsshop\n WHERE player = 'nhanaz'; | Args: []
[11:28:22.838] [Server thread/INFO]: NhanAZ joined the game
[11:28:22.846] [Server thread/CRITICAL]: TypeError: "strpos(): Argument #1 ($haystack) must be of type string, pocketmine\player\Player given" (EXCEPTION) in "D:/Dev/virions/libasynql/src/poggit/libasynql/generic/SqliteStatementImpl" at line 78
--- Stack trace ---
  #0 D:/Dev/virions/libasynql/src/poggit/libasynql/generic/SqliteStatementImpl(78): strpos(object pocketmine\player\Player#30618, string[1] .)
  #1 D:/Dev/virions/libasynql/src/poggit/libasynql/generic/GenericStatementImpl(205): poggit\libasynql\generic\SqliteStatementImpl->formatVariable(object poggit\libasynql\generic\GenericVariable#28779, object pocketmine\player\Player#30618, null, array[0])
  #2 D:/Dev/virions/libasynql/src/poggit/libasynql/base/DataConnectorImpl(200): poggit\libasynql\generic\GenericStatementImpl->format(array[2], null, array[1])
  #3 D:/Dev/virions/libasynql/src/poggit/libasynql/base/DataConnectorImpl(182): poggit\libasynql\base\DataConnectorImpl->executeImpl(string[20] kygektagsshop.insert, array[2], int 2, object Closure#119989, null)
  #4 D:/Dev/virions/libasynql/src/poggit/libasynql/base/DataConnectorImpl(150): poggit\libasynql\base\DataConnectorImpl->executeImplLast(string[20] kygektagsshop.insert, array[2], int 2, object Closure#120143, null)
  #5 plugins/KygekTagsShop/src/Kygekraqmak/KygekTagsShop/TagsActions(241): poggit\libasynql\base\DataConnectorImpl->executeInsert(string[20] kygektagsshop.insert, array[2])
  #6 D:/Dev/virions/libasynql/src/poggit/libasynql/base/DataConnectorImpl(168): Kygekraqmak\KygekTagsShop\TagsActions->Kygekraqmak\KygekTagsShop\{closure}(array[0], array[0])
  #7 D:/Dev/virions/libasynql/src/poggit/libasynql/base/DataConnectorImpl(183): poggit\libasynql\base\DataConnectorImpl::poggit\libasynql\base\{closure}(object poggit\libasynql\result\SqlSelectResult#120010)
  #8 D:/Dev/virions/libasynql/src/poggit/libasynql/base/DataConnectorImpl(213): poggit\libasynql\base\DataConnectorImpl::poggit\libasynql\base\{closure}(array[1])
  #9 D:/Dev/virions/libasynql/src/poggit/libasynql/base/SqlThreadPool(112): poggit\libasynql\base\DataConnectorImpl->poggit\libasynql\base\{closure}(array[1])
  #10 D:/Dev/virions/libasynql/src/poggit/libasynql/base/DataConnectorImpl(296): poggit\libasynql\base\SqlThreadPool->readResults(array[1])
  #11 D:/Dev/virions/libasynql/src/poggit/libasynql/base/SqlThreadPool(65): poggit\libasynql\base\DataConnectorImpl->checkResults()
  #12 pmsrc/vendor/pocketmine/snooze/src/SleeperHandler(123): poggit\libasynql\base\SqlThreadPool->poggit\libasynql\base\{closure}()
  #13 (): --- below is the original stack trace ---()
  #14 D:/Dev/virions/libasynql/src/poggit/libasynql/base/DataConnectorImpl(202): poggit\libasynql\base\DataConnectorImpl->executeImplRaw(array[1], array[1], int 3, object Closure#120041, null)
  #15 D:/Dev/virions/libasynql/src/poggit/libasynql/base/DataConnectorImpl(182): poggit\libasynql\base\DataConnectorImpl->executeImpl(string[17] kygektagsshop.get, array[1], int 3, object Closure#120041, null)
  #16 D:/Dev/virions/libasynql/src/poggit/libasynql/base/DataConnectorImpl(166): poggit\libasynql\base\DataConnectorImpl->executeImplLast(string[17] kygektagsshop.get, array[1], int 3, object Closure#120151, null)
  #17 plugins/KygekTagsShop/src/Kygekraqmak/KygekTagsShop/TagsActions(247): poggit\libasynql\base\DataConnectorImpl->executeSelect(string[17] kygektagsshop.get, array[1], object Closure#120152)
  #18 plugins/KygekTagsShop/src/Kygekraqmak/KygekTagsShop/TagsActions(137): Kygekraqmak\KygekTagsShop\TagsActions->getData(object pocketmine\player\Player#30618, object Closure#120035)
  #19 plugins/KygekTagsShop/src/Kygekraqmak/KygekTagsShop/TagsShop(177): Kygekraqmak\KygekTagsShop\TagsActions->getPlayerTag(object pocketmine\player\Player#30618, object Closure#120035)
  #20 pmsrc/src/event/RegisteredListener(60): Kygekraqmak\KygekTagsShop\TagsShop->onJoin(object pocketmine\event\player\PlayerJoinEvent#120011)
  #21 pmsrc/src/event/Event(62): pocketmine\event\RegisteredListener->callEvent(object pocketmine\event\player\PlayerJoinEvent#120011)
  #22 pmsrc/src/player/Player(834): pocketmine\event\Event->call()
  #23 pmsrc/src/network/mcpe/NetworkSession(710): pocketmine\player\Player->doFirstSpawn()
  #24 pmsrc/src/network/mcpe/NetworkSession(703): pocketmine\network\mcpe\NetworkSession->onClientSpawnResponse()
  #25 pmsrc/src/network/mcpe/handler/SpawnResponsePacketHandler(35): pocketmine\network\mcpe\NetworkSession->pocketmine\network\mcpe\{closure}()
  #26 pmsrc/vendor/pocketmine/bedrock-protocol/src/SetLocalPlayerAsInitializedPacket(42): pocketmine\network\mcpe\handler\SpawnResponsePacketHandler->handleSetLocalPlayerAsInitialized(object pocketmine\network\mcpe\protocol\SetLocalPlayerAsInitializedPacket#119979)
  #27 pmsrc/src/network/mcpe/NetworkSession(394): pocketmine\network\mcpe\protocol\SetLocalPlayerAsInitializedPacket->handle(object pocketmine\network\mcpe\handler\SpawnResponsePacketHandler#102023)
  #28 pmsrc/src/network/mcpe/NetworkSession(357): pocketmine\network\mcpe\NetworkSession->handleDataPacket(object pocketmine\network\mcpe\protocol\SetLocalPlayerAsInitializedPacket#119979, string[2] q.)
  #29 pmsrc/src/network/mcpe/raklib/RakLibInterface(182): pocketmine\network\mcpe\NetworkSession->handleEncoded(string[47] c*d4......KN.4..t....!......@.....LLL. ....{A4.)
  #30 pmsrc/vendor/pocketmine/raklib-ipc/src/RakLibToUserThreadMessageReceiver(42): pocketmine\network\mcpe\raklib\RakLibInterface->onPacketReceive(int 0, string[56] .M.1.$.i.e.x.;........^..3.ze.*...f.z.G.....+..L..?.{K..)
  #31 pmsrc/src/network/mcpe/raklib/RakLibInterface(112): raklib\server\ipc\RakLibToUserThreadMessageReceiver->handle(object pocketmine\network\mcpe\raklib\RakLibInterface#29030)
  #32 pmsrc/vendor/pocketmine/snooze/src/SleeperHandler(123): pocketmine\network\mcpe\raklib\RakLibInterface->pocketmine\network\mcpe\raklib\{closure}()
  #33 pmsrc/vendor/pocketmine/snooze/src/SleeperHandler(82): pocketmine\snooze\SleeperHandler->processNotifications()
  #34 pmsrc/src/Server(1703): pocketmine\snooze\SleeperHandler->sleepUntil(float 1659673702.8513)
  #35 pmsrc/src/Server(1063): pocketmine\Server->tickProcessor()
  #36 pmsrc/src/PocketMine(304): pocketmine\Server->__construct(object BaseClassLoader#2, object pocketmine\utils\MainLogger#3, string[7] D:\Dev\, string[15] D:\Dev\plugins\)
  #37 pmsrc/src/PocketMine(327): pocketmine\server()
  #38 pmsrc(11): require(string[51] phar://D:/Dev/PocketMine-MP.phar/src/PocketMine.php)
--- End of exception information ---
[11:28:22.850] [Server thread/EMERGENCY]: An unrecoverable error has occurred and the server has crashed. Creating a crash dump
[11:28:22.866] [Server thread/EMERGENCY]: Please upload the "D:/Dev/crashdumps/Fri_Aug_5-11.28.22-WIB_2022.log" file to the Crash Archive and submit the link to the Bug Reporting page. Give as much info as you can.
[11:28:22.867] [Server thread/EMERGENCY]: Forcing server shutdown
[11:28:22.868] [Server thread/DEBUG]: Disabling all plugins
```

**Comment by NhanAZ** [comment]
Conflict needs to be resolved
https://github.com/thebigcrafter/KygekTagsShop/pull/16/conflicts
image

**Comment by DavidGlitch04** [comment]
The conflict has been resolved

**Review by NhanAZ — COMMENTED** [review]
Please review the code. Console freezes when a player joins the server.

**Comment by DavidGlitch04** [comment]
Still this issue has been solved

**Comment by NhanAZ** [comment]
```yaml
[12:09:43.957] [Server thread/INFO]: BoxierChimera37 joined the game
[12:09:43.958] [Server thread/DEBUG]: [Player: BoxierChimera37] Exceeded movement rate limit, forcing to last accepted position
[12:09:43.958] [Server thread/DEBUG]: [Player: BlahCoast30765] Exceeded movement rate limit, forcing to last accepted position
[12:09:43.961] [Server thread/DEBUG]: [KygekTagsShop] Queuing mode-2 query: INSERT INTO kygektagsshop (player, tagid)\n VALUES ('blahcoast30765', 0); | Args: []
[12:09:43.963] [Server thread/CRITICAL]: ErrorException: "Undefined array key 0" (EXCEPTION) in "plugins/KygekTagsShop/src/Kygekraqmak/KygekTagsShop/TagsActions" at line 245
--- Stack trace ---
  #0 plugins/KygekTagsShop/src/Kygekraqmak/KygekTagsShop/TagsActions(245): pocketmine\errorhandler\ErrorToExceptionHandler::handle(int 2, string[21] Undefined array key 0, string[74] D:\Dev\plugins\KygekTagsShop\src\Kygekraqmak\KygekTagsShop\TagsActions.php, int 245)
  #1 D:/Dev/virions/libasynql/src/poggit/libasynql/base/DataConnectorImpl(168): Kygekraqmak\KygekTagsShop\TagsActions->Kygekraqmak\KygekTagsShop\{closure}(array[0], array[0])
  #2 D:/Dev/virions/libasynql/src/poggit/libasynql/base/DataConnectorImpl(183): poggit\libasynql\base\DataConnectorImpl::poggit\libasynql\base\{closure}(object poggit\libasynql\result\SqlSelectResult#107289)
  #3 D:/Dev/virions/libasynql/src/poggit/libasynql/base/DataConnectorImpl(213): poggit\libasynql\base\DataConnectorImpl::poggit\libasynql\base\{closure}(array[1])
  #4 D:/Dev/virions/libasynql/src/poggit/libasynql/base/SqlThreadPool(112): poggit\libasynql\base\DataConnectorImpl->poggit\libasynql\base\{closure}(array[1])
  #5 D:/Dev/virions/libasynql/src/poggit/libasynql/base/DataConnectorImpl(296): poggit\libasynql\base\SqlThreadPool->readResults(array[3])
  #6 D:/Dev/virions/libasynql/src/poggit/libasynql/base/SqlThreadPool(65): poggit\libasynql\base\DataConnectorImpl->checkResults()
  #7 pmsrc/vendor/pocketmine/snooze/src/SleeperHandler(123): poggit\libasynql\base\SqlThreadPool->poggit\libasynql\base\{closure}()
  #8 pmsrc/vendor/pocketmine/snooze/src/SleeperHandler(82): pocketmine\snooze\SleeperHandler->processNotifications()
  #9 pmsrc/src/Server(1703): pocketmine\snooze\SleeperHandler->sleepUntil(float 1659676183.9585)
  #10 (): --- below is the original stack trace ---()
  #11 D:/Dev/virions/libasynql/src/poggit/libasynql/base/DataConnectorImpl(202): poggit\libasynql\base\DataConnectorImpl->executeImplRaw(array[1], array[1], int 3, object Closure#121381, null)
  #12 D:/Dev/virions/libasynql/src/poggit/libasynql/base/DataConnectorImpl(182): poggit\libasynql\base\DataConnectorImpl->executeImpl(string[17] kygektagsshop.get, array[1], int 3, object Closure#121381, null)
  #13 D:/Dev/virions/libasynql/src/poggit/libasynql/base/DataConnectorImpl(166): poggit\libasynql\base\DataConnectorImpl->executeImplLast(string[17] kygektagsshop.get, array[1], int 3, object Closure#121399, null)
  #14 plugins/KygekTagsShop/src/Kygekraqmak/KygekTagsShop/TagsActions(246): poggit\libasynql\base\DataConnectorImpl->executeSelect(string[17] kygektagsshop.get, array[1], object Closure#121371)
  #15 plugins/KygekTagsShop/src/Kygekraqmak/KygekTagsShop/TagsActions(137): Kygekraqmak\KygekTagsShop\TagsActions->getData(object pocketmine\player\Player#108294, object Closure#121373)
  #16 plugins/KygekTagsShop/src/Kygekraqmak/KygekTagsShop/TagsShop(177): Kygekraqmak\KygekTagsShop\TagsActions->getPlayerTag(object pocketmine\player\Player#108294, object Closure#121373)
  #17 pmsrc/src/event/RegisteredListener(60): Kygekraqmak\KygekTagsShop\TagsShop->onJoin(object pocketmine\event\player\PlayerJoinEvent#121345)
  #18 pmsrc/src/event/Event(62): pocketmine\event\RegisteredListener->callEvent(object pocketmine\event\player\PlayerJoinEvent#121345)
  #19 pmsrc/src/player/Player(834): pocketmine\event\Event->call()
  #20 pmsrc/src/network/mcpe/NetworkSession(710): pocketmine\player\Player->doFirstSpawn()
  #21 pmsrc/src/network/mcpe/NetworkSession(703): pocketmine\network\mcpe\NetworkSession->onClientSpawnResponse()
  #22 pmsrc/src/network/mcpe/handler/SpawnResponsePacketHandler(35): pocketmine\network\mcpe\NetworkSession->pocketmine\network\mcpe\{closure}()
  #23 pmsrc/vendor/pocketmine/bedrock-protocol/src/SetLocalPlayerAsInitializedPacket(42): pocketmine\network\mcpe\handler\SpawnResponsePacketHandler->handleSetLocalPlayerAsInitialized(object pocketmine\network\mcpe\protocol\SetLocalPlayerAsInitializedPacket#107292)
  #24 pmsrc/src/network/mcpe/NetworkSession(394): pocketmine\network\mcpe\protocol\SetLocalPlayerAsInitializedPacket->handle(object pocketmine\network\mcpe\handler\SpawnResponsePacketHandler#120791)
  #25 plugins/FakePlayer-master/src/muqsit/fakeplayer/DefaultFakePlayerListener(44): pocketmine\network\mcpe\NetworkSession->handleDataPacket(object pocketmine\network\mcpe\protocol\SetLocalPlayerAsInitializedPacket#107292, string[2] q.)
  #26 pmsrc/src/scheduler/ClosureTask(57): muqsit\fakeplayer\DefaultFakePlayerListener::muqsit\fakeplayer\{closure}()
  #27 pmsrc/src/scheduler/TaskHandler(114): pocketmine\scheduler\ClosureTask->onRun()
  #28 pmsrc/src/scheduler/TaskScheduler(141): pocketmine\scheduler\TaskHandler->run()
  #29 pmsrc/src/plugin/PluginManager(498): pocketmine\scheduler\TaskScheduler->mainThreadHeartbeat(int 91)
  #30 pmsrc/src/Server(1811): pocketmine\plugin\PluginManager->tickSchedulers(int 91)
  #31 pmsrc/src/Server(1700): pocketmine\Server->tick()
  #32 pmsrc/src/Server(1063): pocketmine\Server->tickProcessor()
  #33 pmsrc/src/PocketMine(304): pocketmine\Server->__construct(object BaseClassLoader#2, object pocketmine\utils\MainLogger#3, string[7] D:\Dev\, string[15] D:\Dev\plugins\)
  #34 pmsrc/src/PocketMine(327): pocketmine\server()
  #35 pmsrc(11): require(string[51] phar://D:/Dev/PocketMine-MP.phar/src/PocketMine.php)
--- End of exception information ---
[12:09:43.965] [Server thread/EMERGENCY]: An unrecoverable error has occurred and the server has crashed. Creating a crash dump
[12:09:43.981] [Server thread/EMERGENCY]: Please upload the "D:/Dev/crashdumps/Fri_Aug_5-12.09.43-WIB_2022.log" file to the Crash Archive and submit the link to the Bug Reporting page. Give as much info as you can.
[12:09:43.982] [Server thread/EMERGENCY]: Forcing server shutdown
[12:09:43.982] [Server thread/DEBUG]: Disabling all plugins
```

**Comment by DavidGlitch04** [comment]
This bug has been solved

**Comment by NhanAZ** [comment]
## Error: Allows selling `Tag1` infinitely
Reproduce behavior:
- /tagsshop
- Click `Sell Your Tag` button
- Click `Yes` button

Repeat the above behavior many times

**Comment by DavidGlitch04** [comment]
This issue has been solved!

**Comment by NhanAZ** [comment]
## Error: Can't buy `Tag1`

**Comment by DavidGlitch04** [comment]
Fixed!

**Comment by NhanAZ** [comment]
## Error: `Unknown Tag`
Reproduce behavior:
- /tagsshop
- Click `Sell Your Tag`
A form will appear allowing players to sell `Unknown Tag`

**Comment by DavidGlitch04** [comment]
Solved!

**Review by NhanAZ — APPROVED** [review]
Plugin works stably with sqlite database

**Comment by NhanAZ** [comment]
Need to check with mysql database

**Comment by NhanAZ** [comment]
mysql works stably

**Comment by NhanAZ** [comment]
#### TODO: Don't add players to the database if they don't have a Tag

**Comment by NhanAZ** [comment]
```yaml
[18:25:23.160] [Server thread/CRITICAL]: TypeError: "Kygekraqmak\KygekTagsShop\TagsActions::tagExists(): Argument #1 ($tagid) must be of type int, null given, called in D:\Dev\plugins\KygekTagsShop\src\Kygekraqmak\KygekTagsShop\TagsShop.php on line 174" (EXCEPTION) in "plugins/KygekTagsShop/src/Kygekraqmak/KygekTagsShop/TagsActions" at line 124
```

**Comment by DavidGlitch04** [comment]
retry

**Comment by NhanAZ** [comment]
## Error: Can't buy Tag1

**Comment by DavidGlitch04** [comment]
Fixed!

**Review by NhanAZ — APPROVED** [review]
It worked. Nice one.

**Pull request merged**